### ISSUE TITLE
Charge battery up to given maximum

### ISSUE THREAD

**Issue by Biep** [koreader/koreader#5533]
* KOReader version:
* Device: Kobo Aura One - but probably useful for all Kobo devices

#### Issue
I understand battery life is greatly prolonged if the extremes (fully charged or drained) are avoided.  It seems running between 30-70% charged is optimal.

Currently, when 100% charge is reached, after a while the + that indicates charging disappears.  It seems that KOReader can switch the charging off some way or other.  (Or else it just checks whether the percentage has increased in some interval, in which case my proposal is probably impossible.)

It would be great if one could set the percentage at which charging should stop.  There is already the "low battery" warning that helps protect the lower bound; and this would protect the higher bound.
If that is not possible, some warning would be nice, such as blinking the on-off light.

**Comment by NiLuJe** [comment]
AFAIK, charging isn't controlled/controllable via userland, so, this seems technically infeasible.

**Comment by NiLuJe** [comment]
And we've already established in another issue that has been playing with the LED could have unwanted side-effects.

**Comment by Biep** [comment]
Hmm.  And another way to alert the user?  Such as flashing the screen the Start Menu does when installing nickel updates?

**Comment by NiLuJe** [comment]
See the low battery warning feature.

**Comment by NiLuJe** [comment]
Note that, in practice, the charging circuitry should already be doing the brunt work of ensuring sane charging procedures. (i.e., trickle charging when reaching higher charges).

The only case when one might really want to stop at a specific high mark around the low seventies is for long term storage (and/or shipping by air).

**Comment by Biep** [comment]
> See the low battery warning feature.

Yes - that is a good solution for the low value.  I'd like the same for the high value.  It could be settable in the same pop-up, and I suppose to a large extent the same code could be used.
But as one typically doesn't watch the device charge (unless one happens to be reading at the same time), some attention-grabbing warning would be nice.

**Comment by Frenzie** [comment]
Well, fwiw battery life is still pretty good on my 2014 H2O. I'm not exactly sure how often I charge it (and that is a good sign all by itself) but I think it's every 3 weeks or so.

**Comment by Biep** [comment]
For me it would be at most three days, if I am not using it too intensely.
Kobo Aura One from - I think - 2017.
But it depends enormously on *what* I read.  Image pdfs drain it a lot (and so does looking up things in the dictionary).

**Comment by pazos** [comment]
> I understand battery life is greatly prolonged if the extremes (fully charged or drained) are avoided. It seems running between 30-70% charged is optimal.

False, there isn't a single rational argument to stop the charge at 70-80-90%, unless you want to charge again sooner :smiley: . See [TechRepublic's tips for extending lithium battery life](https://www.techrepublic.com/blog/five-apps/five-tips-for-extending-lithium-ion-battery-life/). 

Your device has an integrated circuit that handles the power management and does everything you need. This is: recharge your battery and power you device until 100%, then switch off the battery charge.

It is **safe** the leave the cable plugged for hours, days and weeks without any battery damage.
Your battery will be **heavily damaged** if stays empty during months or years.


> Currently, when 100% charge is reached, after a while the + that indicates charging disappears. It seems that KOReader can switch the charging off some way or other.

We read the information already available (via sysfs). As NiLuJe said it is impossible to stop charging unless the cable is dettached.

> Yes - that is a good solution for the low value. I'd like the same for the high value. It could be settable in the same pop-up, and I suppose to a large extent the same code could be used.
But as one typically doesn't watch the device charge (unless one happens to be reading at the same time), some attention-grabbing warning would be nice.

I would be a completely useless feature and we would contribute to the **high charge problem myth**

**Comment by DJsushi** [comment]
I would like to contribute to this discussion with some useful links:

https://batteryuniversity.com/article/bu-808-how-to-prolong-lithium-based-batteries
https://www.recurrentauto.com/research/one-simple-trick

Charging more often and limiting your charge to 20%-80% will, in fact, increase the lifespan of the battery. I own a Kobo Libra Color, and I want to delay the battery replacement procedure as much as possible, since the replacement will remove the IP protection from my e-reader. So, the limiting of the charging is definitely **not a myth**.

Also, I have found on my Kobo, that there actually IS a way to effectively stop the charging! There is a sysfs interface located at `/sys/class/power_supply/bd71827_bat/charging` that when written a `0` to will effectively stop the charging of the device! So, I am going to try to create a Kobo plugin so that I can set this up in the menu in a nice way :).

**Comment by Frenzie** [comment]
> Charging more often and limiting your charge to 20%-80% will, in fact, increase the lifespan of the battery. I own a Kobo Libra Color, and I want to delay the battery replacement procedure as much as possible, since the replacement will remove the IP protection from my e-reader. So, the limiting of the charging is definitely not a myth.

Mind, what was stated above didn't claim it's false that this is true for the physical battery, but that all of this is already taken care of by the battery itself. We might of course question how well it goes about it, but what presents as 0-100 to the user is something along the lines of 20-80.

> There is a sysfs interface located at /sys/class/power_supply/bd71827_bat/charging that when written a 0 to will effectively stop the charging of the device! 

Huh, nice.

**Comment by DJsushi** [comment]
> Mind, what was stated above didn't claim it's false that this is true for the physical battery, but that all of this is already taken care of by the battery itself. We might of course question how well it goes about it, but what presents as 0-100 to the user is something along the lines of 20-80.

Thanks for the clarification. Well... Yeah. We could question the fact if the battery is really being pushed to the limits by the charging circuitry or not. Maybe I can find that somewhere in the datasheet of the BD71827 controller.
